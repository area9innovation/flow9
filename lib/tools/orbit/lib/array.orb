// Array utility functions library for Orbit

// map: Apply a function to each element in an array
fn map(arr, f) = (
    if length(arr) = 0 then []
    else [f(arr[0])] + map(subrange(arr, 1, length(arr) - 1), f)
);

// filter: Keep only elements that match a predicate
fn filter(arr, pred) = (
    if length(arr) = 0 then []
    else if pred(arr[0]) then 
        [arr[0]] + filter(subrange(arr, 1, length(arr) - 1), pred)
    else 
        filter(subrange(arr, 1, length(arr) - 1), pred)
);

// fold: Accumulate a result based on each array element and an accumulator
fn fold(arr, init, f) = (
    if length(arr) = 0 then init
    else fold(subrange(arr, 1, length(arr) - 1), f(init, arr[0]), f)
);

// mapi: Map with index
fn mapi(arr, f) = (
    fn mapiHelper(arr, f, idx) = (
        if length(arr) = 0 then []
        else [f(arr[0], idx)] + mapiHelper(subrange(arr, 1, length(arr) - 1), f, idx + 1)
    );
    mapiHelper(arr, f, 0)
);

// filteri: Filter with index
fn filteri(arr, pred) = (
    fn filteriHelper(arr, pred, idx) = (
        if length(arr) = 0 then []
        else if pred(arr[0], idx) then 
            [arr[0]] + filteriHelper(subrange(arr, 1, length(arr) - 1), pred, idx + 1)
        else 
            filteriHelper(subrange(arr, 1, length(arr) - 1), pred, idx + 1)
    );
    filteriHelper(arr, pred, 0)
);

// foldi: Fold with index
fn foldi(arr, init, f) = (
    fn foldiHelper(arr, acc, f, idx) = (
        if length(arr) = 0 then acc
        else foldiHelper(subrange(arr, 1, length(arr) - 1), f(acc, arr[0], idx), f, idx + 1)
    );
    foldiHelper(arr, init, f, 0)
);

// tail: Return all elements except the first one
fn tail(arr) = (
    if length(arr) <= 1 then []
    else subrange(arr, 1, length(arr) - 1)
);

// tailFrom: Return all elements from a given index to the end
fn tailFrom(arr, idx) = (
    if idx >= length(arr) then []
    else subrange(arr, idx, length(arr) - idx)
);

// take: Return the first n elements of an array
fn take(arr, n) = (
    if n <= 0 || length(arr) = 0 then []
    else [arr[0]] + take(subrange(arr, 1, length(arr) - 1), n - 1)
);

// iter: Call a function for each element (for side effects)
fn iter(arr, f) = (
    if length(arr) = 0 then 0 // Return 0 instead of ()
    else (
        f(arr[0]);
        iter(subrange(arr, 1, length(arr) - 1), f)
    )
);

// iteri: Iterate with index
fn iteri(arr, f) = (
    fn iteriHelper(arr, f, idx) = (
        if length(arr) = 0 then 0 // Return 0 instead of ()
        else (
            f(arr[0], idx);
            iteriHelper(subrange(arr, 1, length(arr) - 1), f, idx + 1)
        )
    );
    iteriHelper(arr, f, 0)
);

// filtermap: Combine filter and map operations
fn filtermap(arr, f) = (
    if length(arr) = 0 then []
    else (
        let result = f(arr[0]);
        result is (
            Some(val) => [val] + filtermap(subrange(arr, 1, length(arr) - 1), f);
            None() => filtermap(subrange(arr, 1, length(arr) - 1), f);
            _ => filtermap(subrange(arr, 1, length(arr) - 1), f) // For safety
        )
    )
);

// filtermapi: Filtermap with index
fn filtermapi(arr, f) = (
    fn filtermapHelper(arr, f, idx) = (
        if length(arr) = 0 then []
        else (
            let result = f(arr[0], idx);
            result is (
                Some(val) => [val] + filtermapHelper(subrange(arr, 1, length(arr) - 1), f, idx + 1);
                None() => filtermapHelper(subrange(arr, 1, length(arr) - 1), f, idx + 1);
                _ => filtermapHelper(subrange(arr, 1, length(arr) - 1), f, idx + 1) // For safety
            )
        )
    );
    filtermapHelper(arr, f, 0)
);

// removeIndex: Remove an element at a specific index
fn removeIndex(arr, idx) = (
    if idx < 0 || idx >= length(arr) then arr
    else take(arr, idx) + tailFrom(arr, idx + 1)
);

// removeRange: Remove a range of elements
fn removeRange(arr, start, count) = (
    if start < 0 || start >= length(arr) || count <= 0 then arr
    else take(arr, start) + tailFrom(arr, start + count)
);

// removeFirst: Remove the first occurrence of a specific element
fn removeFirst(arr, element) = (
    fn removeFirstHelper(before, after) = (
        if length(after) = 0 then before
        else if after[0] = element then before + subrange(after, 1, length(after) - 1)
        else removeFirstHelper(before + [after[0]], subrange(after, 1, length(after) - 1))
    );
    removeFirstHelper([], arr)
);

// removeAll: Remove all occurrences of a specific element
fn removeAll(arr, element) = (
    filter(arr, \x.x != element != 0)
);

// contains: Check if an element exists in the array
fn contains(arr, element) = (
    if length(arr) = 0 then false
    else if arr[0] = element then true
    else contains(subrange(arr, 1, length(arr) - 1), element)
);

// exists: Check if any element satisfies a predicate
fn exists(arr, pred) = (
    if length(arr) = 0 then false
    else if pred(arr[0]) then true
    else exists(subrange(arr, 1, length(arr) - 1), pred)
);

// forall: Check if all elements satisfy a predicate
fn forall(arr, pred) = (
    if length(arr) = 0 then true
    else if !pred(arr[0]) then false
    else forall(subrange(arr, 1, length(arr) - 1), pred)
);

// uniq: Remove duplicates from an array
fn uniq(arr) = (
    if length(arr) <= 1 then arr
    else (
        fn uniq_helper(remaining, result) = (
            if length(remaining) = 0 then result
            else (
                let current = remaining[0];
                let rest = tail(remaining);
                
                if contains(result, current) then
                    // Skip duplicate
                    uniq_helper(rest, result)
                else
                    // Keep this element
                    uniq_helper(rest, result + [current])
            )
        );
        uniq_helper(arr, [])
    )
);