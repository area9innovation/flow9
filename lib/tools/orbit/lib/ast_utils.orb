// AST utility functions for Orbit

// Create a properly nested AST node by folding over an array of arguments
// This function handles arrays of any length, whereas makeAst expects exactly 2 arguments
fn makeFoldedAst(op, args) = (
    let len = length(args);
    
    if len = 0 then (
        // Handle empty array case - return an appropriate default value
        // This depends on the operator, but we'll return a neutral element
        if op = "&&" then true
        else if op = "||" then false
        else args  // Return empty array as fallback
    ) else if len = 1 then (
        // For single element, just return it without creating an operation
        args[0]
    ) else if len = 2 then (
        // For exactly two elements, we can use makeAst directly
        makeAst(op, args)
    ) else (
        // For 3+ elements, fold to create a proper binary tree
        // Start with the first element and fold over the rest
        let initialAst = args[0];
        let remainingArgs = subrange(args, 1, len - 1);
        
        // Fold over the remaining arguments, combining each with the accumulator
        fold(remainingArgs, initialAst, \acc, element -> (
            makeAst(op, [acc, element])
        ))
    )
);