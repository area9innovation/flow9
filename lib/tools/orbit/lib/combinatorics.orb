// Basic combinatorial functions
fn factorial(n) = (
	if n <= 1 then 1
	else n * factorial(n - 1)
);

fn binomial_coefficient(n, k) = (
	if k < 0 || k > n then 0
	else if k == 0 || k == n then 1
	else (
		// Use multiplicative formula to avoid large intermediate values
		fn binom_helper(i, result) = (
			if i > k then result
			else binom_helper(i + 1, result * (n - k + i) / i)
		);
		binom_helper(1, 1)
	)
);

fn permutations(n, r) = (
	if r < 0 || r > n then 0
	else if r == 0 then 1
	else (
		// Calculate n! / (n-r)!
		fold(range(n-r+1, n+1), 1, \product, i.product * i)
	)
);

// Generate permutations of an array
fn generate_permutations(arr) = (
	if length(arr) <= 1 then [arr]
	else (
		let result = [];
		fn perms_helper(i, perms) = (
			if i >= length(arr) then perms
			else (
				let element = arr[i];
				let rest = remove_at_index(arr, i);
				let sub_perms = generate_permutations(rest);

				let new_perms = fold(sub_perms, perms, \acc, p.acc + [[element] + p]);
				perms_helper(i + 1, new_perms)
			)
		);
		perms_helper(0, [])
	)
);

// Generate combinations of an array
fn generate_combinations(arr, r) = (
	if r > length(arr) then []
	else if r == 0 then [[]]
	else if r == length(arr) then [arr]
	else (
		if length(arr) == 0 then []
		else (
			let first = arr[0];
			let rest = subrange(arr, 1, length(arr) - 1);

			// Combinations that include the first element
			let with_first = map(generate_combinations(rest, r - 1),
													\comb.[first] + comb);

			// Combinations that exclude the first element
			let without_first = generate_combinations(rest, r);

			with_first + without_first
		)
	)
);
