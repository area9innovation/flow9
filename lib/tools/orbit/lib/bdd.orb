import lib/rewrite;
import lib/array;
import lib/logic;

/**
 * Binary Decision Diagrams (BDDs)
 * 
 * This library implements Binary Decision Diagrams (BDDs) for Boolean formulas,
 * based on Shannon's expansion theorem. BDDs provide a canonical representation
 * for Boolean functions, which makes equivalence checking efficient.
 * 
 * BDDs are represented using the if-then-else (Ite) constructor:
 * Ite(var, then_branch, else_branch)
 * 
 * Where:
 * - var is a decision variable
 * - then_branch is the BDD when var is true
 * - else_branch is the BDD when var is false
 * - Terminal nodes are true and false
 */

// ----- VARIABLE ORDERING -----

// Get a numeric value for a variable to determine ordering
fn var_order(var) = (
    if is_not(var) then var_order(get_not_arg(var))
    else var
);

// Check if an expression is a negation
fn is_not(expr) = (
    getConstructor(expr) == "!"
);

// Get the argument of a negation
fn get_not_arg(expr) = (
    expr is (!arg => arg; _ => expr)
);

// ----- BDD CONSTRUCTION -----

// Convert a Boolean formula to a BDD
fn to_bdd(expr) = (
    expr is (
        // Terminal cases
        true => true;
        false => false;
        
        // Negation
		¬e => negate_bdd(to_bdd(e));
        
        // Conjunction (AND)
        a ∧ b => apply_and(to_bdd(a), to_bdd(b));
        
        // Disjunction (OR)
		a ∨ b => apply_or(to_bdd(a), to_bdd(b));
        
        // Already a BDD - ensure consistency
        Ite(v, t, e) => make_ite(v, to_bdd(t), to_bdd(e));
        
        // Any other expression type
        // Variable becomes a simple BDD
        v => if is_var(v) then make_ite(v, true, false) else expr;
    )
);

// Check if an expression is a variable
fn is_var(expr) = (
    let c = getConstructor(expr);
    c != "Ite" && c != "!" && c != "&&" && c != "||" && 
    expr != true && expr != false
);

// Create an Ite node with reduction
fn make_ite(var, then_case, else_case) = (
    // Rule 1: If both branches are equal, eliminate the test
    if then_case == else_case then then_case
    else Ite(var, then_case, else_case)
);

// Negate a BDD
fn negate_bdd(bdd) = (
    bdd is (
        true => false;
        false => true;
        Ite(v, t, e) => make_ite(v, negate_bdd(t), negate_bdd(e));
        _ => !bdd  // For variables or unprocessed expressions
    )
);

// Apply AND operation to two BDDs
fn apply_and(bdd1, bdd2) = (
    // Terminal cases
    if bdd1 == false || bdd2 == false then false
    else if bdd1 == true then bdd2
    else if bdd2 == true then bdd1
    else bdd1 is (
        Ite(v1, t1, e1) => (
            bdd2 is (
                Ite(v2, t2, e2) => (
                    // Choose variable to expand based on ordering
                    let v_order1 = var_order(v1);
                    let v_order2 = var_order(v2);
                    if v_order1 == v_order2 then (
                        // Same variable - combine results
                        make_ite(v1, apply_and(t1, t2), apply_and(e1, e2))
                    ) else if v_order1 < v_order2 then (
                        // v1 comes first - expand on v1
                        make_ite(v1, apply_and(t1, bdd2), apply_and(e1, bdd2))
                    ) else (
                        // v2 comes first - expand on v2
                        make_ite(v2, apply_and(bdd1, t2), apply_and(bdd1, e2))
                    )
                );
                // bdd2 is not an Ite node
                _ => make_ite(v1, apply_and(t1, bdd2), apply_and(e1, bdd2))
            )
        );
        // bdd1 is not an Ite node
        _ => bdd2 is (
            Ite(v2, t2, e2) => make_ite(v2, apply_and(bdd1, t2), apply_and(bdd1, e2));
            _ => bdd1 && bdd2  // Neither is an Ite node
        )
    )
);

// Apply OR operation to two BDDs
fn apply_or(bdd1, bdd2) = (
    // Terminal cases
    if bdd1 == true || bdd2 == true then true
    else if bdd1 == false then bdd2
    else if bdd2 == false then bdd1
    else bdd1 is (
        Ite(v1, t1, e1) => (
            bdd2 is (
                Ite(v2, t2, e2) => (
                    // Choose variable to expand based on ordering
                    let v_order1 = var_order(v1);
                    let v_order2 = var_order(v2);
                    
                    if v_order1 == v_order2 then (
                        // Same variable - combine results
                        make_ite(v1, apply_or(t1, t2), apply_or(e1, e2))
                    ) else if v_order1 < v_order2 then (
                        // v1 comes first - expand on v1
                        make_ite(v1, apply_or(t1, bdd2), apply_or(e1, bdd2))
                    ) else (
                        // v2 comes first - expand on v2
                        make_ite(v2, apply_or(bdd1, t2), apply_or(bdd1, e2))
                    )
                );
                // bdd2 is not an Ite node
                _ => make_ite(v1, apply_or(t1, bdd2), apply_or(e1, bdd2))
            )
        );
        // bdd1 is not an Ite node
        _ => bdd2 is (
            Ite(v2, t2, e2) => make_ite(v2, apply_or(bdd1, t2), apply_or(bdd1, e2));
            _ => bdd1 || bdd2  // Neither is an Ite node
        )
    )
);

// ----- BDD TO LOGIC CONVERSION -----

fn makeOr(a, b) = (
	scatter(`||`([a,b]))
);

fn makeAnd(a, b) = (
	scatter(`&&`([a,b]))
);

fn makeNot(a) = (
	makeAst("!", ([a]))
);

// Convert a BDD back to a standard Boolean formula
fn from_bdd(bdd) = (
    bdd is (
        // Terminal cases
        true => true;
        false => false;
        
        // BDD node
        Ite(v, t, e) => (
            let then_expr = from_bdd(t);
            let else_expr = from_bdd(e);
            
            // Simplify common cases
            if then_expr == true && else_expr == false then v
            else if then_expr == false && else_expr == true then makeNot(v)
            else if then_expr == true then makeOr(v, else_expr)
            else if else_expr == false then makeAnd(v, then_expr)
			else if else_expr == true then makeOr(makeNot(v), then_expr)
            else if then_expr == false then makeAnd(makeNot(v), else_expr)
            else (
				makeOr(
					makeAnd(v, then_expr),
					makeAnd(makeNot(v), else_expr)
				)
			)
        );
        
        // Variable or other expression
        _ => bdd
    )
);

// ----- VARIABLE EXTRACTION -----

// Extract all variables from a logic expression in order
fn extract_variables(expr) = (
    expr is (
        // Terminal cases
        true => [];
        false => [];
        
        // Variable
        v => if is_var(v) then [v] else [];
        
        // Negation
        !e => extract_variables(e);
        
        // AND/OR - combine variables from both sides
        a && b => uniq(sort(extract_variables(a) + extract_variables(b)));
        a || b => uniq(sort(extract_variables(a) + extract_variables(b)));
        
        // BDD node
        Ite(v, t, e) => uniq(sort([v] + extract_variables(t) + extract_variables(e)));
        
        // Fallback
        _ => []
    )
);

// ----- UTILITY FUNCTIONS -----

// Evaluate a BDD with a variable assignment
fn evaluate_bdd(bdd, assignment) = (
    bdd is (
        // Terminal cases
        true => true;
        false => false;
        
        // BDD node - follow appropriate branch based on variable value
        Ite(v, t, e) => (
            let value = assignment(v);
            if value then evaluate_bdd(t, assignment) else evaluate_bdd(e, assignment)
        );
        
        // Variable or unprocessed expression
        _ => assignment(bdd)
    )
);

// ----- EQUIVALENCE CHECKING -----

// Check if two Boolean formulas are equivalent
fn are_equivalent(expr1, expr2) = (
    let bdd1 = to_bdd(expr1);
    let bdd2 = to_bdd(expr2);
    bdd1 == bdd2
);

// ----- QUANTIFICATION OPERATIONS -----

// Variable restriction: substitute variable with a value
fn bdd_restrict(node, var, value) = (
    // Terminal cases are unchanged by restriction
    if node == true || node == false then node
    else node is (
        Ite(v, t, e) => (
            if v == var then
                // Direct substitution when we find the variable
                if value then t else e
            else (
                // For other variables, recurse on both branches
                make_ite(v, bdd_restrict(t, var, value), bdd_restrict(e, var, value))
            )
        );
        // For variables or other expressions
        _ => if node == var then value else node
    )
);

// Existential quantification: ∃v.f
fn bdd_exists(var, node) = (
    // Compute f[v→0] ∨ f[v→1]
    apply_or(
        bdd_restrict(node, var, false),
        bdd_restrict(node, var, true)
    )
);

// Universal quantification: ∀v.f
fn bdd_forall(var, node) = (
    // Compute f[v→0] ∧ f[v→1]
    apply_and(
        bdd_restrict(node, var, false),
        bdd_restrict(node, var, true)
    )
);

// Apply existential quantification for multiple variables
fn bdd_exists_multiple(vars, node) = (
    if length(vars) == 0 then node
    else bdd_exists_multiple(
        tail(vars), 
        bdd_exists(head(vars), node)
    )
);

// Apply universal quantification for multiple variables
fn bdd_forall_multiple(vars, node) = (
    if length(vars) == 0 then node
    else bdd_forall_multiple(
        tail(vars), 
        bdd_forall(head(vars), node)
    )
);

// ----- MODEL COUNTING -----

// Calculate the number of satisfying assignments for a BDD
// Simple implementation without using cache
fn bdd_count_models(node, var_count) = (
    node is (
        // No satisfying assignments for false
        false => 0;
        
        // All remaining variable assignments satisfy true
        true => (
            // 2^remaining_vars
            let result = 1;
            fn calculate_power(base, exponent, acc) = (
                if exponent <= 0 then acc
                else calculate_power(base, exponent - 1, acc * base)
            );
            calculate_power(2, var_count, 1)
        );
        
        // For BDD nodes, compute recursively
        Ite(v, t, e) => (
            // Count satisfying assignments in both branches
            let low_count = bdd_count_models(e, var_count - 1);
            let high_count = bdd_count_models(t, var_count - 1);
            
            // Sum the counts
            low_count + high_count
        );
        
        // For variables or other expressions
        _ => if node then (
            // 2^var_count
            let result = 1;
            fn calculate_power(base, exponent, acc) = (
                if exponent <= 0 then acc
                else calculate_power(base, exponent - 1, acc * base)
            );
            calculate_power(2, var_count, 1)
        ) else 0
    )
);

// Find one satisfying assignment for a BDD
fn bdd_find_satisfying_assignment(node) = (
    if node == false then
        None()  // No satisfying assignment
    else if node == true then
        Some([])  // Empty assignment works
    else node is (
        Ite(v, t, e) => (
            // Try high branch first (v=true)
            let high_result = bdd_find_satisfying_assignment(t);
            high_result is (
                Some(assignment) => Some([Pair(v, true)] + assignment);
                None() => (
                    // Try low branch (v=false)
                    let low_result = bdd_find_satisfying_assignment(e);
                    low_result is (
                        Some(assignment) => Some([Pair(v, false)] + assignment);
                        None() => None()
                    )
                )
            )
        );
        // For variables, assign true
        _ => Some([Pair(node, true)])
    )
);