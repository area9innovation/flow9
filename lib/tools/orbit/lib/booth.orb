import lib/array;

// Linear time Booth's algorithm to find canonical form in cyclic groups
fn booth_canonical(arr) = (
	if length(arr) <= 1 then arr
	else (
		let n = length(arr);

		// Create an auxiliary array which is the original array concatenated with itself
		let double_arr = arr + arr;

		// Initialize failure function array with -1
		// (We use -1 instead of conventional 0 to simplify logic)
		let f = \i. if i < 0 || i >= n then -1 else f(i);

		// Compute the min rotation index in linear time
		fn compute_min_index() = (
			// Initialize starting values
			fn find_min(i, j, k) = (
				// We're done when we've checked all rotations
				if i + k >= 2 * n then j
				else if k >= n then find_min(i + 1, j, 0)
				else (
					// Compute the current indices we're comparing
					let i_idx = (i + k) % (2 * n);
					let j_idx = (j + k) % (2 * n);

					if double_arr[i_idx] < double_arr[j_idx] then
						// i rotation is smaller, j becomes i
						find_min(i, i, k + 1)
					else if double_arr[i_idx] > double_arr[j_idx] then
						// j rotation is smaller, i moves forward
						find_min(i + k + 1, j, 0)
					else
						// Equal so far, continue comparison
						find_min(i, j, k + 1)
				)
			);

			// Start with first two positions
			find_min(0, 0, 0)
		);

		let min_start = compute_min_index();

		// Build the canonical form starting from min_start
		fn build_result(idx, result) = (
			if idx >= n then result
			else build_result(idx + 1, result + [arr[(min_start + idx) % n]])
		);

		build_result(0, [])
	)
);

