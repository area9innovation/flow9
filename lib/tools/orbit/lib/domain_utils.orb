import lib/graph;

// Compare two domains in the hierarchy
// Returns: 
//   -1 if a is more specific than b (a ⊂ ... ⊂ b)
//    0 if a and b are the same
//    1 if b is more specific than a (b ⊂ ... ⊂ a)
//    None() if a and b are not directly related
fn compareDomains(graph, a, b) = (
    if (a == b) then 0  // Same domain
    else if (isReachable(graph, a, b)) then -1  // a is more specific than b
    else if (isReachable(graph, b, a)) then 1   // b is more specific than a
    else None()  // Not directly related
);

// Find the lowest domain from a list of domains
// The lowest domain is the most specific one in the hierarchy
fn findLowestDomain(graph, domains) = (
    if (length(domains) == 0) then None()
    else if (length(domains) == 1) then Some(domains[0])
    else (
        // Start with the first domain as our current lowest
        let initialLowest = domains[0];
        
        // Compare with each other domain to find the lowest
        fold(subrange(domains, 1, length(domains) - 1), initialLowest, \currentLowest, domain -> (
            compareDomains(graph, domain, currentLowest) is (
                -1 => domain;  // This domain is more specific than current lowest
                0 => currentLowest;  // Same domain, keep current
                1 => currentLowest;  // Current lowest is more specific
                None() => (
                    // Domains are not directly related, we keep the current lowest
                    // A more sophisticated approach might look at the full hierarchy
                    currentLowest
                )
            )
        ))
    )
);

// Get the most specific domain a node belongs to
// This is the main function to be used in the OGraph system
fn getNodeMostSpecificDomain(domainGraph, og, nodeId) = (
    // Get all domains the node belongs to
    let domains = getONodeBelongsTo(og, nodeId);
    
    // Find and return the lowest domain from the list
    findLowestDomain(domainGraph, domains)
);

"Domain Hierarchy Utility Library Successfully Loaded"