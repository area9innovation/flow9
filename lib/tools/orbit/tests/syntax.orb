

// Orbit Syntax Test File
// This file tests all syntax in Orbit and is intentionally not semantically meaningful

// 1. IMPORTS
import lib/sort;
import lib/array;

fn quote(a : ast) = a;

let f = quote(

	// 2. COMMENTS
	// Single line comment
	/* Multi-line comment
	spanning multiple lines
	to test block comments */

	// 3. BASIC VALUES AND LITERALS
	let intValue = 42;
	let negativeInt = -123;
	let doubleValue = 3.14159;
	let negativeDouble = -0.123;
	let boolTrue = true;
	let boolFalse = false;
	let stringValue = "Hello, Orbit!";
	let escapeString = "Line 1\nLine 2\tTabbed\"Quoted\"";
	let emptyArray = [];
	let numArray = [1, 2, 3, 4, 5];
	let nestedArray = [[1, 2], [3, 4], [5, 6]];
	let setLiteral = {1, 2, 3, 4};
	let setComprehension = {x | x < 10};
	let ellipsis = ...;

	// 4. ARITHMETIC OPERATORS
	let addition = a + b;
	let subtraction = a - b;
	let multiplication = a * b;
	let division = a / b;
	let modulo = a % b;
	let exponent = a ^ b;
	let negation = -a;
	let multipleOps = (a + b) * (c - d) / e ^ f % g;

	// 5. COMPARISON OPERATORS
	let equals = a == b;
	let notEquals = a != b;
	let lessThan = a < b;
	let lessEquals = a <= b;
	let greaterThan = a > b;
	let greaterEquals = a >= b;

	// 6. LOGICAL OPERATORS
	let logicalAnd = p && q;
	let logicalOr = p || q;
	let logicalNot = !p;
	let complexLogic = (p && q) || (!r && s);

	// 7. VARIABLE BINDINGS
	let singleBinding = 10;
	let multipleBindings = (let x = 5; let y = 10; x + y);
	let nestedBindings = (let outer = (let inner = 42; inner * 2); outer / 2);

	// 8. FUNCTIONS
	// Basic function definition
	fn add(a, b) = a + b;


	// Multi-statement function definition
	fn calculateArea(width, height) = (
		let area = width * height;
		println("Area: " + i2s(area));
		area
	);

	// Function with type annotation
	fn identity(x : int) = x;

	// Function with pattern matching
	fn factorial(n) = (
		if n <= 1 then 1
		else n * factorial(n - 1)
	);

	// 9. LAMBDA EXPRESSIONS
	let incrementLambda = \x.x + 1;
	let multiParamLambda = \(a, b).a * b;
	let lambdaWithEnv = \{captured = 10} x.x + captured;

	// 10. CONDITIONALS
	let simpleIf = if condition then value1 else value2;
	let nestedIf = if a > b then (
		if a > c then a else c
	) else (
		if b > c then b else c
	);
	let multiStatementIf = if condition then (
		let x = 10;
		let y = 20;
		x + y
	) else (
		let z = 30;
		z
	);

	// 11. PATTERN MATCHING
	let simpleMatch = expr is (
		0 => "Zero";
		1 => "One";
		2 => "Two";
		_ => "Other"
	);

	let constructorMatch = shape is (
		Circle(r) => 3.14159 * r * r;
		Rectangle(w, h) => w * h;
		Triangle(b, h) => 0.5 * b * h;
		_ => 0
	);

	let guardedMatch = value is (
		n => "Positive" if n > 0;
		n => "Negative" if n < 0;
		_ => "Zero"
	);

	let nestedMatch = data is (
		Pair(a, b) => a is (
			0 => "First is zero";
			_ => "First is " + i2s(a)
		);
		_ => "Not a pair"
	);

	// 12. CONSTRUCTORS AND FIELD ACCESS
	let point = Point(10, 20);
	let person = Person("John", 30, true);
	let nested = Pair(Point(1, 2), Point(3, 4));
	let fieldAccess = point.x;
	let methodCall = point.distance(origin);

	// 13. SEQUENCES
	let sequence = (expr1; expr2; expr3);
	let nestedSequence = (expr1; (expr2; expr3); expr4);


	// 14. MATH NOTATION
	let setUnion = A ∪ B;
	let setIntersection = A ∩ B;
	let elementOf = x ∈ S;
	let subset = A ⊂ B;

	let forallQuantifier = ∀ x: P(x);
	let existsQuantifier = ∃ x: P(x);

	let logicalAnd2 = p ∧ q;
	let logicalOr2 = p ∨ q;
	let logicalNot2 = ¬p;
	let implies = p ⇒ q;
	let biconditional = p ⇔ q;
	let entailment = A ⊢ B;

	// 15. GREEK LETTERS AND SPECIAL SYMBOLS
	let alpha = α;
	let beta = β;
	let gamma = Γ;
	let pi = π;
	let theta = θ;
	let lambda = λ;
	let omega = Ω;
	let integers = ℤ;
	let rationals = ℚ;
	let reals = ℝ;
	let naturals = ℕ;
	let complexNumbers = ℂ;


	// 16. SUBSCRIPTS AND SUPERSCRIPTS
	let subscripted = x₁;
	let superscripted = x²;
	let subscriptedNegative = x₋₁;
	let complexExponent = e^(i·π);

	// 17. SPECIAL MATH OPERATORS
	let gradient = ∇f;
	let integral = ∫f;
	let summation = ∑f;
	let derivative = d/dx(f);
	let partialDerivative = ∂/∂x(f);
	let tensorProduct = A ⊗ B;
	let directProduct = G × H;
	let composition = f ∘ g;
	let groupOrder = |G|;

	// 18. FUNCTION TYPES AND TYPE OPERATIONS
	let functionType = A → B;
	let typeAnnotation = expr : Type;
	let notTypeAnnotation = expr !: Type;
	let typeSubstitution = expr[T := U];


	// 19. GROUP THEORY
	let normalSubgroup = H ⊲ G;
	let isomorphism = G ≅ H;
	let rightSemiDirectProduct = N ⋊ H;
	let leftSemiDirectProduct = H ⋉ N;

	// 20. REWRITING RULES
	let simpleRule = a + b ⇒ b + a;
	let conditionalRule = expr ⇒ simplified if condition;
	let bidirectionalRule = a * b ⇔ b * a;

	// 21. AST OPERATIONS
	let quotedExpr = quote(a + b * c);
	let makeAstExample = makeAst("+", [a, b]);

	// 22. ENVIRONMENT BINDINGS
	let envBindingLambda = \{x = 10, y = 20} z.x + y + z;


	// 23. MIXED SYNTAX
	let complexExpr = \x.(if x ∈ ℕ then 
		∑{i | i ∈ {1, 2, ..., x}} 
		else 
		∫(λy.y^2)(x));

	// 24. FINAL EXPRESSION
	(let result = 42; println(i2s(result)); result)
);

let g = makeOGraph("graph");
let n = addOGraphWithSub(g, f, []);
setFileContent("test.dot", ograph2dot(g));

let f2 = extractOGraph(g, n);
// println(f2);

println("Second graph");

let g2 = makeOGraph("graph2");
let n2 = addOGraphWithSub(g2, f2, []);
setFileContent("test2.dot", ograph2dot(g2));
