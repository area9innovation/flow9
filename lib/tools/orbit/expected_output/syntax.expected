TODO: Interpret OrSetLiteral: { 1, 2, 3, 4 }
TODO: Interpret OrSetComprehension: { x | (x < 10) }
TODO: Interpret OrEllipsis: ...
Variable not found: a
Variable not found: b
Variable not found: a
Variable not found: b
Variable not found: a
Variable not found: b
Variable not found: a
Variable not found: b
ERROR: Division by zero
Variable not found: a
Variable not found: b
ERROR: Modulo by zero
Variable not found: a
Variable not found: b
Variable not found: a
Variable not found: a
Variable not found: b
Variable not found: c
Variable not found: d
Variable not found: e
Variable not found: f
Variable not found: g
ERROR: Modulo by zero
Variable not found: a
Variable not found: b
Variable not found: a
Variable not found: b
Variable not found: a
Variable not found: b
Variable not found: a
Variable not found: b
Variable not found: a
Variable not found: b
Variable not found: a
Variable not found: b
Variable not found: p
Variable not found: p
Variable not found: q
Variable not found: p
Variable not found: p
Variable not found: r
Variable not found: s
Variable not found: condition
Variable not found: value2
Variable not found: a
Variable not found: b
Variable not found: b
Variable not found: c
Variable not found: c
Variable not found: condition
Variable not found: expr
Variable not found: shape
Variable not found: value
Variable not found: data
TODO: Interpret OrField: point.x
TODO: Interpret OrField: point.distance
Variable not found: origin
Cannot call non-function value: 0
Variable not found: expr1
Variable not found: expr2
Variable not found: expr3
Variable not found: expr1
Variable not found: expr2
Variable not found: expr3
Variable not found: expr4
TODO: Interpret OrUnion: (A ∪ B)
TODO: Interpret OrIntersection: (A ∩ B)
TODO: Interpret OrElementOf: (x ∈ S)
TODO: Interpret OrSubset: (A ⊂ B)
TODO: Interpret OrForall: ∀ (x : P)(x);
let existsQuantifier = ∃ (x : P)(x);
let logicalAnd2 = (p ∧ q);
let logicalOr2 = (p ∨ q);
let logicalNot2 = ¬p;
(let implies = p ⇒ q);
(let biconditional = p ⇔ q);
(let entailment = A ⊢ B);
let alpha = α;
let beta = β;
let gamma = Γ;
let pi = π;
let theta = θ;
let lambda = λ;
let omega = Ω;
let integers = ℤ;
let rationals = ℚ;
let reals = ℝ;
let naturals = ℕ;
let complexNumbers = ℂ;
let subscripted = x₁;
let superscripted = x²;
let subscriptedNegative = x₋₁;
let complexExponent = (e^(i * π));
let gradient = ∇f;
let integral = ∫f;
let summation = ∑f;
let derivative = d/dx f;
let partialDerivative = ∂/∂x f;
let tensorProduct = (A ⊗ B);
let directProduct = (G × H);
let composition = (f ∘ g);
let groupOrder = |G|;
let functionType = (A → B);
let typeAnnotation = (expr : Type);
let notTypeAnnotation = (expr !: Type);
let typeSubstitution = (expr[T := U]);
let normalSubgroup = (H ⊲ G);
let isomorphism = (G ≅ H);
let rightSemiDirectProduct = (N ⋇ H);
let leftSemiDirectProduct = (H ⋆ N);
(let simpleRule = (a + b) ⇒ (b + a));
(let conditionalRule = expr ⇒ simplified condition);
(let bidirectionalRule = (a * b) ⇔ (b * a));
let quotedExpr = quote((a + (b * c)));
let makeAstExample = makeAst("+", [a, b]);
let envBindingLambda = λ{x:10, y:20}z.((x + y) + z);
let complexExpr = λx.if (x ∈ ℕ) then 
	∑{ i | (i ∈ { 1, 2, ..., x }) } 
else 
	∫λy.(y^2)(x);
let result = 42;
println(i2s(result));
result;
0
Result: 0
